package/src/lsif.ts

```diff
@@ -286,6 +287,54 @@ export const noopMaybeProviders = {
 }
 
 export function initLSIF(): MaybeProviders {
+    const provider = (async () => {
+        if (await supportsGraphQL()) {
+            console.log('Using LSIF GraphQL API')
+            return initGraphQL()
+        }
+
+        console.log('Using LSIF HTTP API')
+        return initHTTP()
+    })()
+
+    return {
+        // If graphQL is supported, use the GraphQL implementation.
+        // Otherwise, use the legacy HTTP implementation.
+        definition: async (...args) => (await provider).definition(...args),
+        references: async (...args) => (await provider).references(...args),
+        hover: async (...args) => (await provider).hover(...args),
+    }
+}
+
+async function  supportsGraphQL(): Promise<boolean> {
+    const query = `
+        query SiteVersion {
+            site {
+                productVersion
+            }
+        }
+    `
+
+    const respObj = await queryGraphQL({
+        query,
+        vars:{},
+        sourcegraph,
+    })
+
+    const productVersion:string = respObj.data.site.productVersion
+    if (productVersion === 'dev') {
+        return true
+    }
+
+    // Product versions have the form `49952_2019-12-04_c6f43e0`.
+    // We want to use the GraphQL resolvers if the instance is known to have them.
+    // Check to see if the product version occurs after the date the resolvers became
+    // available. We still support the LSIF HTTP API up to this date, so anything
+    // before it can still fall back safely.
+    return productVersion.split('_')[1].localeCompare(GRAPHQL_API_MINIMUM_DATE) > 0
```

Comment:
`productVersion` could also be something like `3.10.1` https://k8s.sgdev.org/api/console#%7B%22query%22%3A%22query%20%7B%20site%20%7B%20productVersion%20%7D%20%7D%22%7D

Reply:
To be safe, let's run a regex on this to ensure it's the expected format before asserting that the instance supports GraphQL.

Reply:
Yeah I'd like to discuss the possibilities of this a bit more. Quick in-person chat while we're both in SF?

Reply:
Updated to deal with semantic versions as well - this might be a good enough indicator now, but we should still discuss it.